lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,256 @@
+            // closed. On Windows systems this is required for other programs to be able
+            // to open the file for reading.
+            let pub_key_path = pub_key_file.into_temp_path();
+
+            let result = self.run(
+                data,
+                &[
+                    "-Y".as_ref(),
+                    "sign".as_ref(),
+                    "-f".as_ref(),
+                    pub_key_path.as_os_str(),
+                ],
+            );
+
+            Ok(result?)
+        } else {
+            Err(SshError::MissingKey {}.into())
+        }
+    }
+
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<Verification, SignError> {
+        let mut signature_file = tempfile::Builder::new()
+            .prefix(".jj-ssh-sig-")
+            .tempfile()
+            .map_err(SshError::Io)?;
+        signature_file.write_all(signature).map_err(SshError::Io)?;
+        signature_file.flush().map_err(SshError::Io)?;
+
+        let signature_temp_path = signature_file.into_temp_path();
+        let signature_file_path = signature_temp_path.as_os_str();
+
+        let principal = self.find_principal(signature_file_path)?;
+
+        if principal.is_none() {
+            let output = self.run(
+                data,
+                &[
+                    "-Y".as_ref(),
+                    "check-novalidate".as_ref(),
+                    "-s".as_ref(),
+                    signature_file_path,
+                ],
+            );
+
+            return match output {
+                Ok(_) => Ok(Verification::new(
+                    SigStatus::Unknown,
+                    None,
+                    Some("Signature OK. Unknown principal".into()),
+                )),
+                Err(_) => Ok(Verification::new(SigStatus::Bad, None, None)),
+            };
+        }
+
+        match principal {
+            None => {
+                let output = self.run(
+                    data,
+                    &[
+                        "-Y".as_ref(),
+                        "check-novalidate".as_ref(),
+                        "-s".as_ref(),
+                        signature_file_path,
+                    ],
+                );
+
+                match output {
+                    Ok(_) => Ok(Verification::new(SigStatus::Unknown, None, None)),
+                    Err(_) => Ok(Verification::new(SigStatus::Bad, None, None)),
+                }
+            }
+            Some(principal) => {
+                let allowed_signers = self.get_allowed_signers()?;
+
+                let output = self.run(
+                    data,
+                    &[
+                        "-Y".as_ref(),
+                        "verify".as_ref(),
+                        "-s".as_ref(),
+                        signature_file_path,
+                        "-I".as_ref(),
+                        principal.as_ref(),
+                        "-f".as_ref(),
+                        allowed_signers.as_ref(),
+                    ],
```

Comment:
nit: might be better to use the `Command` builder API directly, and extract the post processing part of `run()` to helper function.

`Command::new().arg().arg()` doesn't require all arguments are of certain type.

Reply:
Ok I reworked this into a few different helper functions.